src/java.base/share/classes/jdk/internal/foreign/SegmentFactories.java

```diff
@@ -0,0 +1,211 @@
+/*
+ * Copyright (c) 2023, Oracle and/or its affiliates. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.  Oracle designates this
+ * particular file as subject to the "Classpath" exception as provided
+ * by Oracle in the LICENSE file that accompanied this code.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+package jdk.internal.foreign;
+
+import jdk.internal.access.foreign.UnmapperProxy;
+import jdk.internal.foreign.HeapMemorySegmentImpl.OfByte;
+import jdk.internal.foreign.HeapMemorySegmentImpl.OfChar;
+import jdk.internal.foreign.HeapMemorySegmentImpl.OfDouble;
+import jdk.internal.foreign.HeapMemorySegmentImpl.OfFloat;
+import jdk.internal.foreign.HeapMemorySegmentImpl.OfInt;
+import jdk.internal.foreign.HeapMemorySegmentImpl.OfLong;
+import jdk.internal.foreign.HeapMemorySegmentImpl.OfShort;
+import jdk.internal.misc.Unsafe;
+import jdk.internal.misc.VM;
+import jdk.internal.vm.annotation.ForceInline;
+
+import java.lang.foreign.MemorySegment;
+import java.util.Objects;
+
+/**
+ * This class is used to retrieve concrete memory segment implementations, while making sure that classes
+ * are initialized in the right order (that is, that {@code MemorySegment} is always initialized first).
+ * See {@link SegmentFactories#ensureInitialized()}.
+ */
+public class SegmentFactories {
+
```

Comment:
We could make this a proper utility class by declaring the class final and having a private constructor.